incremental store implementation

tmkasun opened the pull request:
In this PR:

- add incremental store implementation (Nodes, Index, Relations and Property DBs)
- Includes the improvements suggested in https://github.com/miyurud/jasminegraph/pull/104

miyurud commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+    unsigned int edgeRef;
+    unsigned int propRef;
+    char usageBlock;
+    char label[6];
```
Is it possible to move this 6 to a constant? Why do we use 6 here?

tmkasun replied:
It's coming from the definition of Node block structure, We only allow 6 bytes to be stored as node label, Anything longer than that will be stored as a property
![image](https://user-images.githubusercontent.com/3313885/99704712-c445e580-2abe-11eb-808a-341b11b0c2ae.png)

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
I have added another comment, please check.

miyurud commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+        node_block_logger.error("Error while reading prop reference data from block " + std::to_string(blockAddress));
+    }
+
+    if (!NodeBlock::nodesDB->read(&label[0], 6)) {
```
Can we replace 6 with the constant once defined?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/4381d28b31e09cf40c82321c326bdb8aebbe55ce

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+        node_block_logger.error("Error while reading label data from block " + std::to_string(blockAddress));
+    }
+    bool usage = usageBlock == '\1';
+    node_block_logger.debug("Label = "+ std::string(label));
```
Change `"Label = "+ std::string(label)` to `"Label = " + std::string(label)`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+                std::to_string(nodeBlockPointer->addr);
+        }
+    }
+    node_block_logger.debug("Edge ref = "+std::to_string(nodeBlockPointer->edgeRef));
```
Change `"Edge ref = "+std::to_string(nodeBlockPointer->edgeRef)` to `"Edge ref = " + std::to_string(nodeBlockPointer->edgeRef)`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+}
+
+PropertyLink* NodeBlock::getPropertyHead() { return PropertyLink::get(this->propRef); }
+const unsigned long NodeBlock::BLOCK_SIZE = 15;
```
Shall we move this constant declaration to `NodeBlock.h` header file?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeBlock.h:
```diff
@@ -0,0 +1,73 @@
+/**
+Copyright 2020 JasminGraph Team
```
Change `JasminGraph` to `JasmineGraph`

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeBlock.h:
```diff
@@ -0,0 +1,73 @@
+
+#include <cstring>
+#include <fstream>
+#include <iostream>  // for cout
```
I think we are not using cout in these sources anymore. Hence can we remove the line `#include <iostream>`

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeBlock.h:
```diff
@@ -0,0 +1,73 @@
+    std::string id = "";  // Node ID for this block ie: citation paper ID, Facebook accout ID, Twitter account ID etc
+    char usage = false;   // Whether this block is in use or not
+    char label[6] = {
+        0};  // Initialize with null chars label === ID if length(id) < 6 else ID will be store as a Node's property
```
change `will be store` to `will be stored`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+/**
+Copyright 2020 JasminGraph Team
```
Change `JasminGraph` to `JasmineGraph`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+    // TODO (tmkasun): set PropertyLink nextPropertyIndex after validating by modulus check from file number of bytes
+
+    if (dbSize(NodeManager::NODE_DB_PATH) % NodeBlock::BLOCK_SIZE != 0) {
+        node_manager_logger.error("Node DB size does not complies to node block size Path = " +
```
Change `complies` to `comply`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+
+std::unordered_map<std::string, unsigned int> NodeManager::readNodeIndex() {
+    std::ifstream index_db(this->index_db_loc, std::ios::app | std::ios::binary);
+    std::unordered_map<std::string, unsigned int> _nodeIndex;  // temproy node index data holder
```
Change `temproy` to `temporary`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+        int iSize = dbSize(this->index_db_loc);
+        unsigned long dataWidth = NodeManager::INDEX_KEY_SIZE + sizeof(unsigned int);
+        if (iSize % dataWidth != 0) {
+            node_manager_logger.error("Index DB size does not complies to index block size Path = " +
```
Change `complies` to `comply`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

tmkasun commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+            }
+        }
+        return this->setRelationHead(*newRelation);
+        node_block_logger.log("Info: Relocating edge list header", "info");
```
No there is no usage of having an info log there, Will remove it

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Ok, done

tmkasun commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+                    currentRelation = currentRelation->nextDestination();
+                }
+            } else {
+                node_block_logger.warn("Invalid relation block");
```
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

tmkasun commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+}
+
+/**
+ * Return a pointer to maching relation block with the given node if found, Else return NULL
```
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

tmkasun commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -0,0 +1,261 @@
+/**
+Copyright 2020 JasminGraph Team
```
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/1d404d7d8fc79e4d908dcc39667a33b78beeaa26

miyurud replied:
Done

tmkasun commented:
> Why do both lines 38 and 39 carry the same comment?
> 
> `// Size of a node block in bytes`

That was corrected in this commit

https://github.com/miyurud/jasminegraph/pull/115/commits/4381d28b31e09cf40c82321c326bdb8aebbe55ce

miyurud commented:
> > Why do both lines 38 and 39 carry the same comment?
> > `// Size of a node block in bytes`
> 
> That was corrected in this commit
> 
> [4381d28](https://github.com/miyurud/jasminegraph/commit/4381d28b31e09cf40c82321c326bdb8aebbe55ce)

Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+
+NodeManager::NodeManager(std::string mode) {
+    std::ios_base::openmode openMode = std::ios::trunc;  // default is Trunc mode which overrides the entire file
+    if (mode == "app") {
```
Shall we move the literal `"app"` to a constant inside the `NodeManager.h` file?

tmkasun replied:
Yes, We can move this to a static constant, But in header, only static variable definition is allowed. ~[source](https://stackoverflow.com/questions/1563897/static-constant-string-class-member)

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+    // TODO (tmkasun): set PropertyLink nextPropertyIndex after validating by modulus check from file number of bytes
+
+    if (dbSize(NodeManager::NODE_DB_PATH) % NodeBlock::BLOCK_SIZE != 0) {
+        node_manager_logger.error("Node DB size does not comply to node block size Path = " +
```
We should take some action other than logging the error message here. What kind of action can the user take if this error gets thrown?

miyurud replied:
Why don't we throw an error here like done in line 55?

tmkasun replied:
Yes agree, I think we should throw an exception here Since the program can't continue if the data is corrupted

tmkasun replied:
Do we have a JasmineGraph specific exception class ? 
I think it's good to have an exception definition for JasmineGraph.
i:e
```c++
class JasmineException: public exception
{
  virtual const char* what() const throw()
  {
    return "My exception happened";
  }
} jException;

```
~[source](http://www.cplusplus.com/doc/tutorial/exceptions/)
for the time I have used [`std::runtime_error`](https://en.cppreference.com/w/cpp/error/runtime_error) exception

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
I think its good idea to create an Exception class for JasmineGraph. Lets use the name `JasmineGraphException`. I think we can create a top level package called exception and within that host the `JasmineGraphException` class. I will proceed creating this.

miyurud replied:
Created [issue 117](https://github.com/miyurud/jasminegraph/issues/117) to track this. For the moment lets use the approach of throwing an error.

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+        char nodeIDC[NodeManager::INDEX_KEY_SIZE];
+        unsigned int nodeIndexId;
+        for (size_t i = 0; i < iSize / dataWidth; i++) {
+            nodeIDC[NodeManager::INDEX_KEY_SIZE] = {0};  // Fill with null chars before puting data
```
Change `puting` to `putting`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+            _nodeIndex[std::string(nodeIDC)] = nodeIndexId;
+        }
+    } else {
+        node_manager_logger.error("Error while opening the node index DB");
```
why don't we throw an error here?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+            node_manager_logger.error("Something went wrong while adding the new edge/relation source = " +
+                                      std::string(source.id) + " destination = " + std::string(destination.id));
+        }
+
```
remove additional space

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+            destination.updateRelation(newRelation);
+        } else {
+            node_manager_logger.error("Something went wrong while adding the new edge/relation source = " +
+                                      std::string(source.id) + " destination = " + std::string(destination.id));
```
Shouldn't we return -1 here?

tmkasun replied:
Yes I think returning a RelationBlock pointer will solve the issue, In that case, the `null` pointer will indicate an error

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Ok, thanks

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+NodeBlock *NodeManager::addNode(std::string nodeId) {
+    unsigned int assignedNodeIndex;
+    if (this->nodeIndex.find(nodeId) == this->nodeIndex.end()) {
+        node_manager_logger.debug("NodeId not found in index for node ID " + nodeId);
```
This error message does not make any sense. Please revise.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+            source.updateRelation(newRelation);
+            destination.updateRelation(newRelation);
+        } else {
+            node_manager_logger.error("Something went wrong while adding the new edge/relation source = " +
```
Rather than using `Something went wrong` use of `Error` is better. We need to have a concrete error message here. The term `something` indicates we do not know whats going on in the system.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+int NodeManager::dbSize(std::string path) {
+    /*
+        The structure stat contains at least the following members:
+        st_dev     ID of device containing file
```
change `ID of device` to `ID of the device`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+        st_atime   time of last access
+        st_mtime   time of last data modification
+        st_ctime   time of last status change
+        st_blksize a filesystem-specific preferred I/O block size for
```
In this case we should have some example block sizes defined as constants/enumeration.

tmkasun replied:
This comment is actually an extract from [this wiki](https://en.wikipedia.org/wiki/Stat_(system_call)), As a reference for what's in the `stat` structure.
Shall I remove the comment ?

miyurud replied:
We can keep the comment. In addition can you please add the link URL to the code as a comment so that we can track back when needed?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/771d0ebe5f06b0a8fa6801ae04b86f8aa0bc2241

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+ **/
+NodeBlock *NodeManager::get(std::string nodeId) {
+    NodeBlock *nodeBlockPointer = NULL;
+    int debug = std::stoi(nodeId);
```
This variable debug is not used in this function.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+
+    nodeBlockPointer = new NodeBlock(nodeId, blockAddress, propRef, edgeRef, label, usage);
+
+    node_manager_logger.debug("DEBUG: nodeBlockPointer after creating the object edgeRef" +
```
Here `edgeRef"` has to be changed to `edgeRef "`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+    index_db.close();
+}
+
+/**
```
Can we write function level comments for all of the remaining functions in this PR? Having function level comments is best practice although we have not enforced it in this project so far.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.cpp:
```diff
@@ -0,0 +1,254 @@
+    }
+}
+
+const unsigned long NodeManager::INDEX_KEY_SIZE = 8;
```
Can we have these constants defined in a header file?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.h:
```diff
@@ -0,0 +1,53 @@
+    int dbSize(std::string path);
+    void persistNodeIndex();
+    std::unordered_map<std::string, unsigned int> readNodeIndex();
+    static const unsigned long INDEX_KEY_SIZE;  // Size of a index key entry in bytes
```
Change `a index` to `an index`

miyurud replied:
Has to be fixed in lines 30 and 31

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeManager.h:
```diff
@@ -0,0 +1,53 @@
+    static std::string NODE_DB_PATH;  // Size of a index key entry in bytes
+
+   public:
+    static unsigned int nextPropertyIndex; // Next available property block index // unless open in wipe data mode(trunc) need to set this value to property db seekp()/BLOCK_SIZE
```
Reduce the length of this line

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/PropertyLink.cpp:
```diff
@@ -0,0 +1,186 @@
+    if (propertyBlockAddress > 0) {
+        this->propertiesDB->seekg(propertyBlockAddress);
+        char rawName[PropertyLink::MAX_NAME_SIZE] = {0};
+        // char rawValue[PropertyLink::MAX_VALUE_SIZE] = {0};
```
Remove the commented line

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/PropertyLink.cpp:
```diff
@@ -0,0 +1,186 @@
+        if (!this->propertiesDB->read(reinterpret_cast<char*>(&rawName), PropertyLink::MAX_NAME_SIZE)) {
+            property_link_logger.error("Error while reading property name from block " + std::to_string(blockAddress));
+        }
+        property_link_logger.debug("Current FD curser position = " + std::to_string(this->propertiesDB->tellg()) +
```
change `FD` to `file descriptor`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/PropertyLink.cpp:
```diff
@@ -0,0 +1,186 @@
+        }
+
+        this->name = std::string(rawName);
+        // this->value = rawValue;
```
Delete commented line

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/PropertyLink.cpp:
```diff
@@ -0,0 +1,186 @@
+
+/**
+ * pseudo code for inserting an element to a property (single) link list
+ * ### Note: This link list implimentation adds a new link to the tail of the link, unlike in normal case relocating the
```
Change `implimentation` to `implementation`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/PropertyLink.cpp:
```diff
@@ -0,0 +1,186 @@
+ * propertyAddress; If no property address means no property has been stored if(!propertyAddress) { // Means
+ * this is the very first link in the chain yet to be filled store property in this link it self } else if ( //Check
+ * this property name is available in this link) {
+ *      // If so Update the current link
```
Change `Update` to `update`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/PropertyLink.cpp:
```diff
@@ -0,0 +1,186 @@
+    }
+}
+
+/**
```
We should have more descriptive function comment here describing the parameters and the return type

miyurud commented on src/centralstore/incremental/PropertyLink.cpp:
```diff
@@ -0,0 +1,186 @@
+    }
+    return pl;
+}
+const unsigned long PropertyLink::PROPERTY_BLOCK_SIZE =
```
This type of constant definition can be done in a header file.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+bool RelationBlock::setPreviousSource(unsigned int newAddress) {
+    if (this->updateRelationRecords(4, newAddress)) {
+        this->source.preRelationId = newAddress;
+
```
Remove empty line

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

tmkasun commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+/**
+ * Update relation record block given the offset to the recored from the begining, i:e
+ *  recordOffset 0 --> Source address
+ *  recordOffset 1 --> Source address
```
`recordOffset 1 --> Source address` should be `recordOffset 1 --> Destination address`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+}
+
+bool RelationBlock::setNextDestination(unsigned int newAddress) {
+    if (this->updateRelationRecords(6, newAddress)) {
```
Rather than 6 lets use a constant here

tmkasun replied:
Fixed this in https://github.com/miyurud/jasminegraph/pull/115/commits/3ca3404f663a5f12c3ce466721b8fc435447cd26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+}
+
+bool RelationBlock::setPreviousDestination(unsigned int newAddress) {
+    if (this->updateRelationRecords(8, newAddress)) {
```
Rather than 8 lets use a constant here

tmkasun replied:
Fixed this in https://github.com/miyurud/jasminegraph/pull/115/commits/3ca3404f663a5f12c3ce466721b8fc435447cd26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+ *  recordOffset 0 --> Source address
+ *  recordOffset 1 --> Source address
+ *  recordOffset 2 --> Source's next relation block address
+ *  recordOffset 3 --> Source's next relation block partition id
```
Better if we could document the rest of the recordOffset values (4, 5, ...,9) as well here

tmkasun replied:
Fixed this in https://github.com/miyurud/jasminegraph/pull/115/commits/3ca3404f663a5f12c3ce466721b8fc435447cd26

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+        } else {
+            throw "Error occurred while adding a new property link to " + std::to_string(this->addr) + " node block";
+        }
+
```
Lets remove empty line

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+    PropertyLink* current = this->getPropertyHead();
+    while (current) {
+        allProperties.insert({current->name, current->value});
+        PropertyLink* _temp = current->next();
```
Better not have `_` here because AFAIK we have not defined variables starting with `_` in JasmineGraph.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.cpp:
```diff
@@ -0,0 +1,332 @@
+}
+
+const unsigned long RelationBlock::BLOCK_SIZE = RelationBlock::RECORD_SIZE * 11;
+// One relation block holds 11 recods such as source addres, destination address, source next relation address ect . .
```
change `ect` to `etc.`

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.h:
```diff
@@ -0,0 +1,72 @@
+#include <cstring>
```
We need to put license header here

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/RelationBlock.h:
```diff
@@ -0,0 +1,72 @@
+        : addr(addr), source(source), destination(destination), propertyAddress(propertyAddress){};
+
+    char usage;
+    unsigned int addr;  // Block size * block ID ID for this block
```
why `ID ID` repeated?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud commented on src/util/logger/Logger.cpp:
```diff
@@ -32,8 +33,11 @@ void Logger::log(std::string message, const std::string log_type) {
+        logger->debug(message);
     } else {
-        //std::cout << "Invalid logging type" << std::endl;
+        // std::cout << "Invalid logging type" << std::endl;
```
Lets remove this commented line.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9

miyurud replied:
Done

miyurud reviewed (changes requested):
I have added more comments. Please address them one by one.

tmkasun commented:
> I have added more comments. Please address them one by one.

Thanks for reviewing, I have addressed all the review comments and fixed them in the following commits

- https://github.com/miyurud/jasminegraph/pull/115/commits/ed31ba0af33e9629173808960eef3cfe90c5dbe9
- https://github.com/miyurud/jasminegraph/pull/115/commits/3ca3404f663a5f12c3ce466721b8fc435447cd26
